### ISSUE TITLE
Bug: [no-unnecessary-template-expression] False positive suggesting `${T}` -> `T`

### ISSUE THREAD

**Issue by ashvinsrivatsa** [typescript-eslint/typescript-eslint#10738]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.4&fileType=.tsx&code=C4TwDgpgBAKgjAHgIJQgD2BAdgEwM5R7ABOAllgOYB8UAvFAAYAkA3kgL4MDcAsAFChIsAEzJUGbPkIly1OlCS8%2B-bAFcAtlACiUFvyhQAhvIBEAMwD2FkwBp9UAEamHh4rf7t%2B-ADYRgUNDgALlhELRp6cysTJV9-NGEQmFFw00trLiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK18iALbF4QifQAexaIMocm6KImjQO0SODABfEEaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type T1<A extends string> = `${A}`; // <- linter suggests simplifying this
type T2<A extends string> = A;

enum E {
  a = "foo",
  b = "bar",
}

let x1: T1<E> = "foo";
let x2: T2<E> = "foo"; // <- compiler error, if we were to use the linter's suggested simplification
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-template-expression": "error"
  },
};
```

### tsconfig

```jsonc
{}
```

### Expected Result

The ["no-unnecessary-template-expression"](https://typescript-eslint.io/rules/no-unnecessary-template-expression/) rule suggests simplifying this:

```ts
type T<A extends string> = `${A}`;
```

To this:

```ts
type T<A extends string> = A;
```

But because these types are not equivalent, as demonstrated in the playground repro, it should not recommend this.

### Actual Result

The rule suggests the simplification mentioned above.

### Additional Info

Regarding why one might even want a type like this, see https://stackoverflow.com/q/52393730/. tl;dr it lets you flatten TypeScript enums-of-strings into a union of their string values.

**Comment by kirkwaiblinger** [comment]
Hi @ashvinsrivatsa ! 

I think you're right and we should treat this as a bug. The rule does already have special-casing for enum types, but it only applies for concrete enums. In this case, the rule sees `'${A}'` as though `A` were replaced with its constraint, i.e. as `'${string}'` (a common pattern in our rules), which does become `string`, whereas `'${A}'` is not converted to `A` even if `A extends string` (presumably because of the enum case). I'm thinking therefore that the appropriate resolution would be simply to exempt all type parameters from the check.

Thanks for reporting this, and for the simplicity of the issue writeup and minimal reproduction 🙂  

> Regarding why one might even want a type like this...

Heh - this question came up for me in the PR too 😅: https://github.com/typescript-eslint/typescript-eslint/pull/10612#discussion_r1926105462

**Comment by yeonjuan** [comment]
let me fix it! :)
